File: skills/openclix-init/templates/flutter/core/openclix.dart
import '../engine/trigger_service.dart';
import '../infrastructure/flutter_device_locale_provider.dart';
import '../models/openclix_types.dart';
import '../services/config_loader.dart';
import '../services/config_validator.dart';
import '../services/language_resolver.dart';
import '../services/utils.dart';

class OpenClixDependencies {
  final OpenClixLocalMessageScheduler messageScheduler;
  final CampaignStateRepositoryPort campaignStateRepository;
  final OpenClixClock? clock;
  final OpenClixLifecycleStateReader? lifecycleStateReader;
  final OpenClixLogger? logger;

  const OpenClixDependencies({
    required this.messageScheduler,
    required this.campaignStateRepository,
    this.clock,
    this.lifecycleStateReader,
    this.logger,
  });
}

class DefaultClock implements OpenClixClock {
  @override
  String now() => DateTime.now().toUtc().toIso8601String();
}

class DefaultLifecycleStateReader implements OpenClixLifecycleStateReader {
  String currentAppState = 'foreground';

  @override
  String getAppState() => currentAppState;

  @override
  void setAppState(String state) {
    currentAppState = state;
  }

  @override
  void dispose() {}
}

class DefaultLogger implements OpenClixLogger {
  OpenClixLogLevel logLevel;

  static const Map<OpenClixLogLevel, int> logLevelOrder = {
    OpenClixLogLevel.debug: 0,
    OpenClixLogLevel.info: 1,
    OpenClixLogLevel.warn: 2,
    OpenClixLogLevel.error: 3,
    OpenClixLogLevel.none: 4,
  };

  DefaultLogger(this.logLevel);

  @override
  void setLogLevel(OpenClixLogLevel level) {
    logLevel = level;
  }

  @override
  void debug(String message, [Object? argument]) {
    if (shouldLog(OpenClixLogLevel.debug)) {
      print(
        '[OpenClix] DEBUG: $message${argument != null ? ' $argument' : ''}',
      );
    }
  }

  @override
  void info(String message, [Object? argument]) {
    if (shouldLog(OpenClixLogLevel.info)) {
      print('[OpenClix] INFO: $message${argument != null ? ' $argument' : ''}');
    }
  }

  @override
  void warn(String message, [Object? argument]) {
    if (shouldLog(OpenClixLogLevel.warn)) {
      print('[OpenClix] WARN: $message${argument != null ? ' $argument' : ''}');
    }
  }

  @override
  void error(String message, [Object? argument]) {
    if (shouldLog(OpenClixLogLevel.error)) {
      print(
        '[OpenClix] ERROR: $message${argument != null ? ' $argument' : ''}',
      );
    }
  }

  bool shouldLog(OpenClixLogLevel targetLevel) {
    return logLevelOrder[targetLevel]! >= logLevelOrder[logLevel]!;
  }
}

bool isRemoteEndpoint(String endpoint) {
  return endpoint.startsWith('http://') || endpoint.startsWith('https://');
}

const int maximumEventLogSize = 5000;

class OpenClix {
  static OpenClixConfig? config;
  static TriggerService? triggerService;
  static bool initialized = false;
  static CampaignStateRepositoryPort? campaignStateRepository;
  static OpenClixLocalMessageScheduler? messageScheduler;
  static OpenClixClock? clock;
  static OpenClixLifecycleStateReader? lifecycleStateReader;
  static OpenClixLogger? logger;
  static OpenClixDependencies? dependencies;
  static LanguageResolver? _languageResolver;

  OpenClix._();

  static Future<void> initialize(
    OpenClixConfig config,
    OpenClixDependencies dependencies,
  ) async {
    if (initialized) {
      throw StateError(
        'OpenClix is already initialized. Call OpenClix.reset() before re-initializing.',
      );
    }

    OpenClix.config = config;
    OpenClix.dependencies = dependencies;
    messageScheduler = dependencies.messageScheduler;
    campaignStateRepository = dependencies.campaignStateRepository;
    clock = dependencies.clock ?? DefaultClock();
    lifecycleStateReader =
        dependencies.lifecycleStateReader ?? DefaultLifecycleStateReader();
    logger = dependencies.logger ?? DefaultLogger(config.logLevel);
    logger?.setLogLevel(config.logLevel);

    _languageResolver = LanguageResolver(
      sdkDefaultLanguage: config.defaultLanguage,
      deviceLocaleProvider: FlutterDeviceLocaleProvider(),
    );

    triggerService = TriggerService(createTriggerServiceDependencies());

    logger?.info('Initializing OpenClix SDK...');

    if (isRemoteEndpoint(config.endpoint)) {
      try {
        final requestHeaders = <String, String>{
          ...(config.extraHeaders ?? const <String, String>{}),
        };

        if (config.projectId != null && config.projectId!.isNotEmpty) {
          requestHeaders['x-openclix-project-id'] = config.projectId!;
        }

        if (config.apiKey != null && config.apiKey!.isNotEmpty) {
          requestHeaders['x-openclix-api-key'] = config.apiKey!;
        }

        final loadedConfig = await loadConfig(
          config.endpoint,
          options: ConfigLoaderOptions(
            headers: requestHeaders,
            timeoutMs: config.sessionTimeoutMs,
          ),
        );

        final validationResult = validateConfig(loadedConfig);

        if (validationResult.valid) {
          for (final warning in validationResult.warnings) {
            logger?.warn(
              'Config validation warning [${warning.code}]: ${warning.message}',
            );
          }

          triggerService?.replaceConfig(loadedConfig);

          try {
            await evaluateInternal('app_boot');
          } catch (evaluationError) {
            logger?.warn(
              'Initial app_boot evaluation failed:',
              evaluationError,
            );
          }

          logger?.info(
            'Config loaded successfully '
            '(version: ${loadedConfig.configVersion}, '
            'campaigns: ${loadedConfig.campaigns.length})',
          );
        } else {
          for (final error in validationResult.errors) {
            logger?.error(
              'Config validation error [${error.code}]: ${error.message}',
            );
          }
          logger?.warn(
            'Config validation failed. SDK initialized without campaign config.',
          );
        }
      } catch (loadError) {
        logger?.warn(
          'Failed to load config from endpoint. SDK initialized without campaign config. '
          'Use OpenClixCampaignManager.replaceConfig() to set config manually.',
          loadError,
        );
      }
    } else {
      logger?.info(
        'Non-HTTP endpoint provided. '
        'Use OpenClixCampaignManager.replaceConfig() to set campaign config.',
      );
    }

    initialized = true;
    logger?.info('OpenClix SDK initialized successfully.');
  }

  static Future<void> trackEvent(
    String name, [
    Map<String, JsonValue>? properties,
  ]) async {
    assertInitialized();

    final event = Event(
      id: generateUUID(),
      name: name,
      sourceType: EventSourceType.app,
      properties: properties,
      createdAt: clock!.now(),
    );

    await persistEvent(event);
    logger?.debug('Event tracked: $name');

    try {
      await evaluateInternal('event_tracked', event);
    } catch (evaluationError) {
      logger?.warn("Evaluation after event '$name' failed:", evaluationError);
    }
  }

  static Future<void> trackSystemEvent(
    SystemEventName name, [
    Map<String, JsonValue>? properties,
  ]) async {
    assertInitialized();

    final event = Event(
      id: generateUUID(),
      name: name.value,
      sourceType: EventSourceType.system,
      properties: properties,
      createdAt: clock!.now(),
    );

    await persistEvent(event);
  }

  static Future<void> handleNotificationDelivered(
    Map<String, Object?> payload,
  ) async {
    assertInitialized();

    await trackSystemEvent(
      SystemEventName.messageDelivered,
      compactProperties({
        'campaign_id': extractString(payload, const [
          'campaignId',
          'campaign_id',
        ]),
        'queued_message_id': extractString(payload, const [
          'queuedMessageId',
          'queued_message_id',
        ]),
        'channel_type':
            extractString(payload, const ['channelType', 'channel_type']) ??
            ChannelType.appPush.value,
      }),
    );
  }

  static Future<String?> handleNotificationOpened(
    Map<String, Object?> payload,
  ) async {
    assertInitialized();

    final landingUrl = extractString(payload, const [
      'landingUrl',
      'landing_url',
    ]);

    await trackSystemEvent(
      SystemEventName.messageOpened,
      compactProperties({
        'campaign_id': extractString(payload, const [
          'campaignId',
          'campaign_id',
        ]),
        'queued_message_id': extractString(payload, const [
          'queuedMessageId',
          'queued_message_id',
        ]),
        'channel_type':
            extractString(payload, const ['channelType', 'channel_type']) ??
            ChannelType.appPush.value,
        'landing_url': landingUrl,
      }),
    );

    return landingUrl;
  }

  static Future<void> reset() async {
    final loggerAtResetStart = logger;

    if (campaignStateRepository != null) {
      try {
        await campaignStateRepository!.clearCampaignState();
      } catch (error) {
        loggerAtResetStart?.warn(
          'Failed to clear campaign state during reset:',
          error,
        );
      }

      try {
        await campaignStateRepository!.clearEvents();
      } catch (error) {
        loggerAtResetStart?.warn(
          'Failed to clear event log during reset:',
          error,
        );
      }
    }

    if (messageScheduler != null) {
      try {
        final pendingMessages = await messageScheduler!.listPending();
        for (final pendingMessage in pendingMessages) {
          await messageScheduler!.cancel(pendingMessage.id);
        }
      } catch (error) {
        loggerAtResetStart?.warn(
          'Failed to clear scheduled messages during reset:',
          error,
        );
      }
    }

    try {
      lifecycleStateReader?.dispose();
    } catch (error) {
      loggerAtResetStart?.warn(
        'Failed to dispose lifecycle state reader during reset:',
        error,
      );
    }

    config = null;
    triggerService = null;
    initialized = false;
    campaignStateRepository = null;
    messageScheduler = null;
    clock = null;
    lifecycleStateReader = null;
    logger = null;
    dependencies = null;
    _languageResolver = null;

    loggerAtResetStart?.info('OpenClix SDK reset complete.');
  }

  static void setLogLevel(OpenClixLogLevel level) {
    logger?.setLogLevel(level);
  }

  static void setLanguage(String languageCode) {
    assertInitialized();
    _languageResolver?.setLanguage(languageCode);
  }

  static String? getLanguage() {
    assertInitialized();
    return _languageResolver?.getLanguage();
  }

  static void clearLanguage() {
    assertInitialized();
    _languageResolver?.clearLanguage();
  }

  static void handleAppForeground() {
    if (!initialized) return;

    lifecycleStateReader?.setAppState('foreground');
    logger?.debug('App entered foreground');

    evaluateInternal('app_foreground').catchError((Object error) {
      logger?.warn('app_foreground evaluation failed:', error);
      return null;
    });
  }

  static TriggerService? getTriggerServiceInternal() => triggerService;

  static OpenClixClock? getClockInternal() => clock;

  static OpenClixLogger? getLoggerInternal() => logger;

  static CampaignStateRepositoryPort? getCampaignStateRepositoryInternal() {
    return campaignStateRepository;
  }

  static OpenClixLocalMessageScheduler? getMessageSchedulerInternal() {
    return messageScheduler;
  }

  static bool isInitializedInternal() => initialized;

  static void assertInitialized() {
    if (!initialized) {
      throw StateError(
        'OpenClix is not initialized. Call OpenClix.initialize() before using the SDK.',
      );
    }
  }

  static TriggerServiceDependencies createTriggerServiceDependencies() {
    return TriggerServiceDependencies(
      campaignStateRepository: campaignStateRepository!,
      messageScheduler: messageScheduler!,
      clock: clock!,
      logger: logger!,
      recordEvent: (event) async {
        await persistEvent(event);
      },
      languageResolver: _languageResolver,
    );
  }

  static Future<void> persistEvent(Event event) async {
    if (campaignStateRepository == null) {
      logger?.debug(
        "Event store is not available; skipping persistence for event '${event.name}'.",
      );
      return;
    }

    try {
      await campaignStateRepository!.appendEvents([event], maximumEventLogSize);
    } catch (error) {
      logger?.warn("Failed to persist event '${event.name}':", error);
    }
  }

  static String? extractString(Map<String, Object?> source, List<String> keys) {
    for (final key in keys) {
      final value = source[key];
      if (value is String && value.isNotEmpty) {
        return value;
      }
    }
    return null;
  }

  static Map<String, JsonValue> compactProperties(
    Map<String, JsonValue?> values,
  ) {
    final compacted = <String, JsonValue>{};
    values.forEach((key, value) {
      if (value != null) {
        compacted[key] = value;
      }
    });
    return compacted;
  }

  static Future<TriggerResult?> evaluateInternal(
    String trigger, [
    Event? event,
  ]) async {
    if (triggerService == null) {
      return null;
    }

    return triggerService!.trigger(
      TriggerContext(trigger: trigger, event: event, now: clock?.now()),
    );
  }
}

File: skills/openclix-init/templates/ios/Core/OpenClix.swift
import Foundation

private final class DefaultClock: OpenClixClock, Sendable {
    func now() -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.string(from: Date())
    }
}

private final class DefaultLifecycleReader: OpenClixLifecycleStateReader, @unchecked Sendable {
    private let lock = NSLock()
    private var appState: String = "foreground"

    func getAppState() -> String {
        lock.lock()
        let state = appState
        lock.unlock()
        return state
    }

    func setAppState(_ newState: String) {
        lock.lock()
        appState = newState
        lock.unlock()
    }
}

private final class DefaultLogger: OpenClixLogger, @unchecked Sendable {
    private static let logLevelOrder: [OpenClixLogLevel: Int] = [
        .debug: 0,
        .info: 1,
        .warn: 2,
        .error: 3,
        .none: 4,
    ]

    private let lock = NSLock()
    private var level: OpenClixLogLevel

    init(level: OpenClixLogLevel) {
        self.level = level
    }

    func setLogLevel(_ level: OpenClixLogLevel) {
        lock.lock()
        defer { lock.unlock() }
        self.level = level
    }

    func debug(_ message: String, _ args: Any...) {
        guard shouldLog(.debug) else { return }
        print("[OpenClix:DEBUG] \(message)", args.isEmpty ? "" : " \(args)")
    }

    func info(_ message: String, _ args: Any...) {
        guard shouldLog(.info) else { return }
        print("[OpenClix:INFO] \(message)", args.isEmpty ? "" : " \(args)")
    }

    func warn(_ message: String, _ args: Any...) {
        guard shouldLog(.warn) else { return }
        print("[OpenClix:WARN] \(message)", args.isEmpty ? "" : " \(args)")
    }

    func error(_ message: String, _ args: Any...) {
        guard shouldLog(.error) else { return }
        print("[OpenClix:ERROR] \(message)", args.isEmpty ? "" : " \(args)")
    }

    private func shouldLog(_ targetLevel: OpenClixLogLevel) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        let currentLevel = level
        let currentOrder = DefaultLogger.logLevelOrder[currentLevel] ?? 4
        let targetOrder = DefaultLogger.logLevelOrder[targetLevel] ?? 0
        return targetOrder >= currentOrder
    }
}

private func isRemoteEndpoint(_ endpoint: String) -> Bool {
    return endpoint.hasPrefix("http://") || endpoint.hasPrefix("https://")
}

private let maximumEventLogSize = 5_000

public final class OpenClix {

    actor Coordinator {
        private var config: OpenClixConfig?
        private var triggerService: TriggerService?
        private var initialized = false

        private var campaignStateRepository: OpenClixCampaignStateRepository?
        private var messageScheduler: OpenClixMessageScheduler?
        private var clock: OpenClixClock?
        private var lifecycleReader: DefaultLifecycleReader?
        private var logger: DefaultLogger?
        private var languageResolver: LanguageResolver?

        func initialize(config: OpenClixConfig) async {
            guard !initialized else {
                logger?.warn("OpenClix is already initialized. Call OpenClix.reset() before re-initializing.")
                return
            }

            self.config = config

            let campaignStateRepository = FileCampaignStateRepository()
            let messageScheduler = LocalNotificationScheduler()
            let clock = DefaultClock()
            let lifecycleReader = DefaultLifecycleReader()
            let logger = DefaultLogger(level: config.logLevel)

            self.campaignStateRepository = campaignStateRepository
            self.messageScheduler = messageScheduler
            self.clock = clock
            self.lifecycleReader = lifecycleReader
            self.logger = logger

            let languageResolver = LanguageResolver(
                sdkDefaultLanguage: config.defaultLanguage,
                deviceLocaleProvider: IOSDeviceLocaleProvider()
            )
            self.languageResolver = languageResolver

            logger.info("Initializing OpenClix SDK...")

            let triggerService = TriggerService(
                dependencies: TriggerServiceDependencies(
                    campaignStateRepository: campaignStateRepository,
                    messageScheduler: messageScheduler,
                    clock: clock,
                    logger: logger,
                    languageResolver: languageResolver
                )
            )
            self.triggerService = triggerService

            if isRemoteEndpoint(config.endpoint) {
                do {
                    var requestHeaders = config.extraHeaders ?? [:]
                    if let projectId = config.projectId {
                        requestHeaders["x-openclix-project-id"] = projectId
                    }
                    if let apiKey = config.apiKey {
                        requestHeaders["x-openclix-api-key"] = apiKey
                    }

                    let timeoutSeconds = config.sessionTimeoutMs.map {
                        TimeInterval(Double($0) / 1000.0)
                    }

                    if let loadedConfig = try await ConfigLoader.load(
                        endpoint: config.endpoint,
                        timeoutSeconds: timeoutSeconds,
                        extraHeaders: requestHeaders
                    ) {
                        let validationResult = validateConfig(loadedConfig)

                        if validationResult.valid {
                            for warning in validationResult.warnings {
                                logger.warn(
                                    "Config validation warning [\(warning.code)]: \(warning.message)"
                                )
                            }

                            await triggerService.replaceConfig(loadedConfig)

                            let appBootContext = TriggerContext(
                                trigger: .app_boot,
                                now: clock.now()
                            )
                            _ = await triggerService.trigger(appBootContext)

                            logger.info(
                                "Config loaded successfully (version: \(loadedConfig.config_version), campaigns: \(loadedConfig.campaigns.count))"
                            )
                        } else {
                            for error in validationResult.errors {
                                logger.error(
                                    "Config validation error [\(error.code)]: \(error.message)"
                                )
                            }
                            logger.warn(
                                "Config validation failed. SDK initialized without campaign config."
                            )
                        }
                    }
                } catch {
                    logger.warn(
                        "Failed to load config from endpoint. SDK initialized without campaign config. "
                        + "Use OpenClixCampaignManager.replaceConfig() to set config manually. "
                        + "Error: \(error.localizedDescription)"
                    )
                }
            } else {
                logger.info(
                    "Non-HTTP endpoint provided. Use OpenClixCampaignManager.replaceConfig() to set campaign config."
                )
            }

            initialized = true
            logger.info("OpenClix SDK initialized successfully.")
        }

        func trackEvent(_ name: String, properties: [String: JsonValue]?) async {
            assertInitialized()

            guard let clock = clock,
                  let triggerService = triggerService else {
                return
            }

            let event = Event(
                id: generateUUID(),
                name: name,
                source_type: .app,
                properties: properties,
                created_at: clock.now()
            )

            await persistEvent(event)
            logger?.debug("Event tracked: \(name)")

            let context = TriggerContext(
                event: event,
                trigger: .event_tracked,
                now: clock.now()
            )
            _ = await triggerService.trigger(context)
        }

        func trackSystemEvent(_ name: SystemEventName, properties: [String: JsonValue]?) async {
            assertInitialized()

            guard let clock = clock else { return }

            let event = Event(
                id: generateUUID(),
                name: name.rawValue,
                source_type: .system,
                properties: properties,
                created_at: clock.now()
            )

            await persistEvent(event)
        }

        func handleNotificationDelivered(payload: [String: Any]) async {
            assertInitialized()

            let campaignId = extractString(from: payload, keys: "campaignId", "campaign_id")
            let queuedMessageId = extractString(
                from: payload,
                keys: "queuedMessageId",
                "queued_message_id"
            )
            let channelType = extractString(from: payload, keys: "channelType", "channel_type")
                ?? ChannelType.app_push.rawValue

            await trackSystemEvent(
                .openClixMessageDelivered,
                properties: compactProperties(
                    [
                        "campaign_id": campaignId.map { JsonValue.string($0) },
                        "queued_message_id": queuedMessageId.map { JsonValue.string($0) },
                        "channel_type": .string(channelType),
                    ]
                )
            )
        }

        func handleNotificationOpened(payload: [String: Any]) async -> String? {
            assertInitialized()

            let campaignId = extractString(from: payload, keys: "campaignId", "campaign_id")
            let queuedMessageId = extractString(
                from: payload,
                keys: "queuedMessageId",
                "queued_message_id"
            )
            let channelType = extractString(from: payload, keys: "channelType", "channel_type")
                ?? ChannelType.app_push.rawValue
            let landingUrl = extractString(from: payload, keys: "landingUrl", "landing_url")

            await trackSystemEvent(
                .openClixMessageOpened,
                properties: compactProperties(
                    [
                        "campaign_id": campaignId.map { JsonValue.string($0) },
                        "queued_message_id": queuedMessageId.map { JsonValue.string($0) },
                        "channel_type": .string(channelType),
                        "landing_url": landingUrl.map { JsonValue.string($0) },
                    ]
                )
            )

            return landingUrl
        }

        func reset() async {
            let previousLogger = logger

            if let campaignStateRepository = campaignStateRepository {
                do {
                    try await campaignStateRepository.clearCampaignState()
                } catch {
                    previousLogger?.warn(
                        "Failed to clear campaign state during reset: \(error.localizedDescription)"
                    )
                }

                do {
                    try await campaignStateRepository.clearEvents()
                } catch {
                    previousLogger?.warn(
                        "Failed to clear event log during reset: \(error.localizedDescription)"
                    )
                }
            }

            if let messageScheduler = messageScheduler {
                do {
                    let pendingMessages = try await messageScheduler.listPending()
                    for pendingMessage in pendingMessages {
                        try await messageScheduler.cancel(pendingMessage.id)
                    }
                } catch {
                    previousLogger?.warn(
                        "Failed to clear scheduled messages during reset: \(error.localizedDescription)"
                    )
                }
            }

            config = nil
            triggerService = nil
            initialized = false
            campaignStateRepository = nil
            messageScheduler = nil
            clock = nil
            lifecycleReader = nil
            logger = nil
            languageResolver = nil

            previousLogger?.info("OpenClix SDK reset complete.")
        }

        func setLogLevel(_ level: OpenClixLogLevel) async {
            logger?.setLogLevel(level)
            await triggerService?.setLogLevel(level)
        }

        func handleAppForeground() async {
            guard initialized else { return }

            lifecycleReader?.setAppState("foreground")
            logger?.debug("App entered foreground")

            if let triggerService = triggerService,
               let clock = clock {
                _ = await triggerService.trigger(
                    TriggerContext(
                        trigger: .app_foreground,
                        now: clock.now()
                    )
                )
            }
        }

        func setLanguage(_ languageCode: String) {
            languageResolver?.setLanguage(languageCode)
        }

        func getLanguage() -> String? {
            return languageResolver?.getLanguage()
        }

        func clearLanguage() {
            languageResolver?.clearLanguage()
        }

        func getTriggerService() -> TriggerService? { return triggerService }
        func getClock() -> OpenClixClock? { return clock }
        func getLogger() -> OpenClixLogger? { return logger }
        func getCampaignStateRepository() -> OpenClixCampaignStateRepository? {
            return campaignStateRepository
        }
        func getMessageScheduler() -> OpenClixMessageScheduler? { return messageScheduler }
        func isInitialized() -> Bool { return initialized }

        private func persistEvent(_ event: Event) async {
            guard let campaignStateRepository = campaignStateRepository else {
                logger?.debug(
                    "Event store is not available; skipping persistence for event '\(event.name)'."
                )
                return
            }

            do {
                try await campaignStateRepository.appendEvents(
                    [event],
                    maxEntries: maximumEventLogSize
                )
            } catch {
                logger?.warn(
                    "Failed to persist event '\(event.name)': \(error.localizedDescription)"
                )
            }
        }

        private func extractString(from source: [String: Any], keys: String...) -> String? {
            for key in keys {
                guard let value = source[key] as? String else { continue }
                if !value.isEmpty {
                    return value
                }
            }
            return nil
        }

        private func compactProperties(
            _ values: [String: JsonValue?]
        ) -> [String: JsonValue] {
            var compacted: [String: JsonValue] = [:]
            for (key, value) in values {
                if let value {
                    compacted[key] = value
                }
            }
            return compacted
        }

        private func assertInitialized() {
            precondition(
                initialized,
                "OpenClix is not initialized. Call OpenClix.initialize() before using the SDK."
            )
        }
    }

    static let coordinator = Coordinator()

    private init() {}

    public static func initialize(config: OpenClixConfig) async {
        await coordinator.initialize(config: config)
    }

    public static func trackEvent(
        _ name: String,
        properties: [String: JsonValue]? = nil
    ) async {
        await coordinator.trackEvent(name, properties: properties)
    }

    public static func trackSystemEvent(
        name: SystemEventName,
        properties: [String: JsonValue]? = nil
    ) async {
        await coordinator.trackSystemEvent(name, properties: properties)
    }

    public static func handleNotificationDelivered(payload: [String: Any]) async {
        await coordinator.handleNotificationDelivered(payload: payload)
    }

    public static func handleNotificationOpened(payload: [String: Any]) async -> String? {
        return await coordinator.handleNotificationOpened(payload: payload)
    }

    public static func reset() async {
        await coordinator.reset()
    }

    public static func setLogLevel(_ level: OpenClixLogLevel) {
        Task {
            await coordinator.setLogLevel(level)
        }
    }

    public static func handleAppForeground() {
        Task {
            await coordinator.handleAppForeground()
        }
    }

    public static func setLanguage(_ languageCode: String) async {
        await coordinator.setLanguage(languageCode)
    }

    public static func getLanguage() async -> String? {
        return await coordinator.getLanguage()
    }

    public static func clearLanguage() async {
        await coordinator.clearLanguage()
    }
}
